gh-1820 revise create_index() page (#1831)
tarantool/doc: doc/reference/reference_lua/box_space/create_index.rst

    .. _box_space-create_index:

    ..  method:: create_index(index-name [, options ])

        Create an :ref:`index <index-box_index>`.

        It is mandatory to create an index for a space before trying to insert

>         tuples into it, or select tuples from it. The first created index, which
>         will be used as the primary-key index, must be unique.

Reviewer: Can we straighten the logic a bit?

```suggestion
        tuples into it, or select tuples from it. The first created index
        will be used as the primary-key index, so it must be unique.
```

        :return: index object
        :rtype:  index_object

        **Possible errors:**

        * too many parts;
        * index '...' already exists;
        * primary key must be unique.

        Building or rebuilding a large index will cause occasional
        :ref:`yields <atomic-cooperative_multitasking>`
        so that other requests will not be blocked.

>         If the other requests cause an illegal situation such as a duplicate key

Reviewer: Is "illegal situation" a term? Do we define it anywhere?

>         in a unique index, the index building or rebuilding will fail.

Reviewer: ```suggestion
        in a unique index, building or rebuilding such index will fail.
```

            |                     | faster, ``false`` -- an index size is reduced by half |                                  |                               |
            +---------------------+-------------------------------------------------------+----------------------------------+-------------------------------+

        The options in the above chart are also applicable for
        :doc:`/reference/reference_lua/box_index/alter`.

>         **Note re storage engine:** vinyl has extra options which by default are

Reviewer: "re" is not so popular :)

```suggestion
        **Note about storage engine:** vinyl has extra options which by default are
```

Author: We have this phrase declared here https://www.tarantool.io/en/doc/latest/book/box/engines/
So if you want to change the wording, let's do it in another issue.

Reviewer: Hmm, okay.

        The current values can be seen by selecting from
        :doc:`/reference/reference_lua/box_space/_index`.

        **Example:**

>         ..  code-block:: tarantoolsession

Reviewer: nice )

            tarantool> s=box.schema.space.create('tester')
            ---
            ...
            tarantool> s:create_index('primary', {unique = true, parts = {
                     > {field = 1, type = 'unsigned'},
                     > {field = 2, type = 'string'}
                     > }})
            ---
            - unique: true
              parts:
              - type: unsigned
                is_nullable: false
                fieldno: 1
              - type: string
                is_nullable: false
                fieldno: 2
              id: 0
              space_id: 512
              type: TREE
              name: primary
            ...

.. _details_about_index_field_types:

.. _box_space-index_field_types:

--------------------------------------------------------------------------------
Details about index field types
--------------------------------------------------------------------------------

Index field types differ depending on what values are allowed,
and what index types are allowed.

..  container:: table stackcolumn

    ..  rst-class:: left-align-column-1
    ..  rst-class:: left-align-column-2

>     ..  list-table::

Reviewer: A table is cool )

        :widths: 10 45 20 15
        :header-rows: 1

        *   - Index field type
            - What can be it it
            - Where it is legal
            - Examples

>         *   - **unsigned**

Reviewer: Looks like code to me, should be formatted as code.

```suggestion
        *   - ``unsigned``
```

            - null (input with ``msgpack.NULL`` or ``yaml.NULL`` or ``json.NULL``),
              booleans (true or false), or
              integers between -9223372036854775808 and 18446744073709551615, or
              single-precision floating point numbers, or
              double-precision floating-point numbers, or
              exact numbers, or
              strings, or
              (varbinary) byte arrays.
              When there is a mix of types, the key order is: null,
              then booleans, then numbers, then strings, then byte arrays
            - memtx TREE or HASH indexes;

              vinyl TREE indexes
            - null

              true

              -1

              1.234

              ''

              'ру'

        *   - **nil**
            - Additionally, `nil` is allowed with any index field type
              if :ref:`is_nullable=true <box_space-is_nullable>` is specified
            -
            -

.. _box_space-is_nullable:

--------------------------------------------------------------------------------
Allowing null for an indexed key
--------------------------------------------------------------------------------

> If the index type is TREE, and the index is not the primary index,

Reviewer: TREE should be formatted as inline code.

then the ``parts={...}`` clause may include ``is_nullable=true`` or
``is_nullable=false`` (the default). If ``is_nullable`` is true,

> then it is legal to insert ``nil`` or an equivalent such as ``msgpack.NULL``
> (or it is legal to insert nothing at all for trailing nullable fields).

Reviewer: parentheses → new sentence

```suggestion
then it is legal to insert ``nil`` or an equivalent such as ``msgpack.NULL``.
It is also legal to insert nothing at all when using trailing nullable fields.
```

    format = {{'id', 'unsigned'}, {'data', 'map'}}
    s:format(format)
    parts = {{'data.FIO["firstname"]', 'str'}, {'data.FIO["surname"]', 'str'}}
    i = s:create_index('info', {parts = parts})
    s:insert({1, {FIO={firstname='Ahmed', surname='Xi'}}})

**Note re storage engine:** vinyl supports only the TREE index type, and vinyl
secondary indexes must be created before tuples are inserted.

.. _box_space-path_multikey:

--------------------------------------------------------------------------------
Using the path option with [*]
--------------------------------------------------------------------------------

The string in a path option can contain '[*]' which is called
an array index placeholder. Indexes defined with this are useful
for JSON documents that all have the same structure.

For example, when creating an index on field#2 for a string document
that will start with ``{'data': [{'name': '...'}, {'name': '...'}]``,
the parts section in the create_index request could look like:
``parts = {{field = 2, type = 'str', path = 'data[*].name'}}``.
Then tuples containing names can be retrieved quickly with
``index_object:select({key-value})``.

> In fact a single field can have multiple keys, as in this example

Reviewer: 1. In fact, a single field (comma)
2. The sentence is too complex. Try breaking it into two or three separate sentences.

  x.component
* '[*]' must not appear in the path of a primary-key
* if an index has ``unique=true`` and has a path with '[*]'
  then duplicate keys from different tuples are disallowed but duplicate keys
  for the same tuple are allowed
* as with :ref:`Using the path option for map fields <box_space-path>`,
  the field's value must have the structure that the path definition implies,
  or be nil (nil is not indexed)

.. _box_space-index_func:

--------------------------------------------------------------------------------
Making a functional index with space_object:create_index()
--------------------------------------------------------------------------------

Functional indexes are indexes that call a user-defined function for forming
the index key, rather than depending entirely on the Tarantool default formation.
Functional indexes are useful for condensing or truncating or reversing or
any other way that users want to customize the index.

The function definition must expect a tuple (which has the contents of
fields at the time a data-change request happens) and must return a tuple
(which has the contents that will actually be put in the index).

> The space must have a memtx engine.

Reviewer: This section has a lot of one-sentence paragraphs. Maybe they can be joined into a few longer paragraphs or grouped in a list.

Reviewer: > в [*]-секции хотели сделать маркированный список, но не получилось, надо исправить

The function must be :ref:`persistent <box_schema-func_create_with-body>`
and deterministic.

The key parts must not depend on JSON paths.

The ``create_index`` definition must include specification of all key parts,
and the function must return a table which has the same number of key parts
with the same types.

The function must access key-part values by index, not by field name.

Functional indexes must not be primary-key indexes.

Functional indexes cannot be altered and the function cannot be changed if
it is used for an index, so the only way to change them is to drop the index
and create it again.

Only sandboxed functions are suitable for functional indexes.

> **Example:**

Reviewer: Why bold and not a heading?

A function could make a key using only the first letter of a string field.

#.  Make the space: the space needs a primary-key field, which is not
    the field that we will use for the functional index.

    ..  code-block:: lua

        box.schema.space.create('x', {engine = 'memtx'})
        box.space.x:create_index('i',{parts={{field = 1, type = 'string'}}})

#.  Make the function: the function expects a tuple. In this example it will
    work on tuple[2] because the key source is field number 2 in what we will

>     insert. Use string.sub() from the string module to get the first character.

Reviewer: ```suggestion
    insert. Use ``string.sub()`` from the ``string`` module to get the first character:
```

    ..  code-block:: lua

        lua_code = [[function(tuple) return {string.sub(tuple[2],1,1)} end]]

> #.  Make the function persistent: use the box.schema.func.create function for this.

Reviewer: Either formatting or a link to the document.
```suggestion
#.  Make the function persistent using the ``box.schema.func.create`` function:
```

    ..  code-block:: lua

        box.schema.func.create('F',
            {body = lua_code, is_deterministic = true, is_sandboxed = true})

> #.  Make the functional index: specify the fields whose values will be passed

Reviewer: This is the first mention of this index, so we should use "a": "Make a functional index".

>     to the function. Specify the function.

Reviewer: We're giving an example after this sentence.

```suggestion
    to the function. Specify the function:
```

    ..  code-block:: lua

        box.space.x:create_index('j',{parts={{field = 1, type = 'string'}},func = 'F'})

#.  Test. Insert a few tuples. Select using only the first letter, it will work
    because that is the key. Or, select using the same function as was used for

>     insertion

Reviewer: ```suggestion
    insertion:
```

        - - ['a', 'wombat']
        ...

Here is the full code of the example:

..  code-block:: lua

    box.schema.space.create('x', {engine = 'memtx'})
    box.space.x:create_index('i',{parts={{field = 1, type = 'string'}}})
    lua_code = [[function(tuple) return {string.sub(tuple[2],1,1)} end]]
    box.schema.func.create('F',
        {body = lua_code, is_deterministic = true, is_sandboxed = true})
    box.space.x:create_index('j',{parts={{field = 1, type = 'string'}},func = 'F'})
    box.space.x:insert{'a', 'wombat'}
    box.space.x:insert{'b', 'rabbit'}
    box.space.x.index.j:select('w')
    box.space.x.index.j:select(box.func.F:call({{'x', 'wombat'}}));

> Functions for functional indexes can return **multiple keys**. Such functions are

Reviewer: Why bold?
